core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -84,94 +119,105 @@ public static void stopThreadPool() {
         THREAD_POOL = null;
     }
 
-    public void testNoMaster() {
+    public void testNoMaster() throws InterruptedException {
         final String index = "test";
 
         clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
 
-        DiscoveryNodes.Builder builder = DiscoveryNodes.builder(clusterService.state().nodes());
-        builder.masterNodeId(null);
-        clusterService.setState(ClusterState.builder(clusterService.state()).nodes(builder));
+        DiscoveryNodes.Builder noMasterBuilder = DiscoveryNodes.builder(clusterService.state().nodes());
+        noMasterBuilder.masterNodeId(null);
+        clusterService.setState(ClusterState.builder(clusterService.state()).nodes(noMasterBuilder));
 
         String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
 
+        CountDownLatch latch = new CountDownLatch(1);
         AtomicBoolean noMaster = new AtomicBoolean();
-        assert !noMaster.get();
+        AtomicBoolean retried = new AtomicBoolean();
+        AtomicBoolean success = new AtomicBoolean();
 
-        shardStateAction.shardFailed(clusterService.state(), getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
-            @Override
-            public void onShardFailedNoMaster() {
-                noMaster.set(true);
-            }
+        setUpMasterRetryVerification(noMaster, retried, latch);
 
+        shardStateAction.shardFailed(getRandomShardRouting(index), indexUUID, "test", getSimulatedFailure(), new ShardStateAction.Listener() {
             @Override
-            public void onShardFailedFailure(DiscoveryNode master, TransportException e) {
-
+            public void onSuccess() {
+                success.set(true);
+                latch.countDown();
             }
         });
 
+        latch.await();
+
         assertTrue(noMaster.get());
+        assertTrue(retried.get());
+        assertTrue(success.get());
     }
 
-    public void testFailure() {
+    public void testMasterChannelException() throws InterruptedException {
         final String index = "test";
 
         clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
 
         String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
 
-        AtomicBoolean failure = new AtomicBoolean();
-        assert !failure.get();
+        CountDownLatch latch = new CountDownLatch(1);
+        AtomicBoolean noMaster = new AtomicBoolean();
+        AtomicBoolean retried = new AtomicBoolean();
```

Comment:
can we make this an AtomicLong and also test that the number of times we retried is what we expect?